Implement a generic read_csv method

titoeb opened the pull request:
Looking into implementing #66.

Mec-iS commented:
hey. Thanks again for contributing.

Some notes:
* take a look to the warnings and linting, always run `cargo fmt --all -- --check` and `cargo clippy --all-features -- -Drust-2018-idioms -Dwarnings`.
* always add a dosctring on top of a new module using `//!` when you explain the content and objectives for the module: in this case for example
```
//! Load a `Matrix` from CSV file
//! * ...
//! * only accepting commas as separators
//! ...
 ```
 The best examples of documentation are in files like `src/linalg/cholesky.rs`, we would like to stick to that standard.
 * Adding the `FromStr` trait to `RealNumber` is for sure the rustucean way of doing it 👍🏼 maybe adding a test for `RealNumber::from_str` in `src/math/num.rs` would be an additional control, the more tests the better
 * The convention for generic type in Rust and in this crate is `T` (not `Type`)
 
 Great work, especially with the test coverage and error handling 💯

titoeb commented:
> hey. Thanks again for contributing.
> 
> Some notes:
> 
> * take a look to the warnings and linting, always run `cargo fmt --all -- --check` and `cargo clippy --all-features -- -Drust-2018-idioms -Dwarnings`.
> * always add a dosctring on top of a new module using `//!` when you explain the content and objectives for the module: in this case for example
> 
> ```
> //! Load a `Matrix` from CSV file
> //! * ...
> //! * only accepting commas as separators
> //! ...
> //! Example:
> //! ```
> //! ...
> //! matrix_from_csv::<f64, Vec<_>, DenseMatrix<_>>(
>                     test_csv_file
>                         .expect("Temporary file could not be written.")
>                         .path()
>                 )
> //!
> ```
> 
> "Example" part is important as code snippets are also compiled and used as tests. The best examples of documentation are in files like `src/linalg/cholesky.rs`, we would like to stick to that standard.
> 
> * Adding the `FromStr` trait to `RealNumber` is for sure the rustucean way of doing it 👍🏼 maybe adding a test for `RealNumber::from_str` in `src/math/num.rs` would be an additional control, the more tests the better
> * The convention for generic type in Rust and in this crate is `T` (not `Type`)
> 
> Great work, especially with the test coverage and error handling 💯

Hi @Mec-iS !
Thanks for already looking into this, your comments make sense 👍 This is still a Draft PR so there is still a lot of work for me to do until it is ready to opened. Feel free to continue to comment on the progress of course, but also take into consideration that the implementation will probably change quite alot until then. I mainly opened the PR to make visible that I work on the issue and give you already a sense of the direction in which I am heading, I am very happy that you are overall like the changes.

titoeb commented:
> hey. Thanks again for contributing.
> 
> Some notes:
> 
> * take a look to the warnings and linting, always run `cargo fmt --all -- --check` and `cargo clippy --all-features -- -Drust-2018-idioms -Dwarnings`.
> * always add a dosctring on top of a new module using `//!` when you explain the content and objectives for the module: in this case for example
> 
> ```
> //! Load a `Matrix` from CSV file
> //! * ...
> //! * only accepting commas as separators
> //! ...
> //! Example:
> //! ```
> //! ...
> //! matrix_from_csv::<f64, Vec<_>, DenseMatrix<_>>(
>                     test_csv_file
>                         .expect("Temporary file could not be written.")
>                         .path()
>                 )
> //!
> ```
> 
> "Example" part is important as code snippets are also compiled and used as tests. The best examples of documentation are in files like `src/linalg/cholesky.rs`, we would like to stick to that standard.
> 
> * Adding the `FromStr` trait to `RealNumber` is for sure the rustucean way of doing it 👍🏼 maybe adding a test for `RealNumber::from_str` in `src/math/num.rs` would be an additional control, the more tests the better
> * The convention for generic type in Rust and in this crate is `T` (not `Type`)
> 
> Great work, especially with the test coverage and error handling 100

- Linting now runs through
- I added docstrings when it was applicable.
- I replaced `Type` with `T`.
